app/javascript/mastodon/actions/compose.js

```diff
@@ -212,17 +214,43 @@ export function clearComposeSuggestions() {
 };
 
 export function fetchComposeSuggestions(token) {
-  return (dispatch, getState) => {
-    api(getState).get('/api/v1/accounts/search', {
-      params: {
-        q: token,
-        resolve: false,
-        limit: 4,
-      },
-    }).then(response => {
-      dispatch(readyComposeSuggestions(token, response.data));
-    });
-  };
+  let leading = token[0];
+
+  if (leading === '@') {
+    // handle search
+    return (dispatch, getState) => {
+      api(getState).get('/api/v1/accounts/search', {
+        params: {
+          q: token.slice(1), // remove the '@'
+          resolve: false,
+          limit: 4,
+        },
+      }).then(response => {
+        dispatch(readyComposeSuggestions(token, response.data));
+      });
+    };
+  } else if (leading === ':') {
+    // shortcode
+    let allShortcodes = Object.keys(emojione.emojioneList);
+    // TODO when we have custom emojons merged, add them to this shortcode list
+    return (dispatch) => {
+      dispatch(readyComposeSuggestionsTxt(token, allShortcodes.filter((sc) => {
+        return sc.indexOf(token) === 0;
```

Comment:
how performant is this in the normal case? fast enough that the auto-complete keeps up with you typing? i know we've done a lot of work to make emojis more performant in matching. Maybe using the trie makes sense here? That doesn't allow for matches in the middle of the string though....

Reply:
I didn't see any lag when testing it. Don't forget it's all debounced at 500 ms, same as account search

Reply:
oof :( 500 ms is a ton of latency, especially for locally stored values.

Reply:
try it, it doesn't feel bad at all

Reply:
hmmm okay it could be a bit faster.

I don't think we can lower the delay without some more aggressive changes tho - it's here: https://github.com/glitch-soc/mastodon/blob/master/app/javascript/mastodon/features/compose/components/compose_form.js#L89

Reply:
nvm lol